server/src/test/java/io/crate/auth/HttpAuthUpstreamHandlerTest.java

```diff
@@ -272,4 +281,67 @@ public void testUnauthorizedUserWithDisabledHBA() throws Exception {
         assertThat(handler.authorized()).isFalse();
         assertUnauthorized(ch.readOutbound(), "trust authentication failed for user \"Aladdin\"\n");
     }
+
+    @Test
+    public void test_user_authentication_with_jwt_token() throws Exception {
+        Roles roles = () -> List.of(JWT_USER);
+        Authentication authentication = mock(Authentication.class);
+        AuthenticationMethod jwtAuth = mock(JWTAuthenticationMethod.class);
+        when(authentication.resolveAuthenticationType(eq(JWT_USER.name()), any(ConnectionProperties.class)))
+            .thenReturn(jwtAuth);
+        when(jwtAuth.authenticate(any(Credentials.class),any(ConnectionProperties.class))).thenReturn(JWT_USER);
```

Comment:
Now that the class is parameterized it should be possible to use a real instance here instead of a mock. 

Probably the same for `Authentication`. It should be possible to use `HostBasedAuthentication` with settings that fit, and only mock the `DnsResolver`

Same for the other tests.

Reply:
`test_user_authentication_with_jwt_token_user_not_found` doesn't use mocks,
`test_user_authentication_with_jwt_token_verified_per_request` needs mocking as we calculate number of times, function was called.

Regarding this test:even though `AuthenticationMethod` is parameterized, we still need mock its parameters, especially jwkProvider, like it's done in `UserAuthenticationMethodTest`. `HostBasedAuthentication` would be also mock because by default it creates JwtAuth Method with `JWTAuthenticationMethod::jwkProvider`. 

 We would essentially replace one bunch of mocks with another, not sure it worth it.

Reply:
If we don't make use of the parameterization, then we can remove it again.  The way it is right now it doesn't make any sense.

> We would essentially replace one bunch of mocks with another, not sure it worth it.

There's a difference between mocking a third party dependency, and mocking _our_ classes, which we want to cover as much as possible.

From what I can tell to make the jwkProvider parameterization actually useful we'd have to extend it to the `HostBasedAuthentication` constructor, and then use it somewhat like this:

```
        Roles roles = () -> List.of(JWT_USER);
        Settings settings = Settings.builder()
            .put("auth.host_based.enabled", true)
            .put("auth.host_based.config.0.user", "John")
            .put("auth.host_based.config.0.method", "jwt")
            .put("auth.host_based.config.0.protocol", "http")
            .build();
        JwkProvider jwkProvider = new JwkProvider() {

            @Override
            public Jwk get(String keyId) throws JwkException {
                return mock(Jwk.class, Answers.RETURNS_MOCKS); // or real Jwk object
            }
        };
        Authentication authentication = new HostBasedAuthentication(
            settings,
            roles,
            mock(DnsResolver.class),
            url -> jwkProvider
        );
```


Not insisting on that, but please don't do a half compromise of adding something without using it properly.

Reply:
Okay taking that back, I missed the `UserAuthenticationMethodTest` test cases.
But I'm getting the impression that some of these tests are redundant?

Reply:
Checked again, tests cover different non-happy path scenarios, have different error messages.

Somewhat similar but in different components are
`HttpAuthUpstreamHandlerTest.test_user_authentication_with_jwt_token_user_not_found`

and 
`UserAuthenticationMethodTest.test_jwt_authentication_user_not_found_throws_an_error`

but they are covering how different components react to not found user.